### PR TITLE
Add Yearly Date Rule

### LINKED ISSUE #7988: Yearly Date Rule
#### Expected Behavior

`self.date_rules.year_start` exists

#### Actual Behavior

The longest period for date rules is currently monthly.

#### Potential Solution

N/A

#### Reproducing the Problem

N/A

#### System Information

N/A

#### Checklist

- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

- [ ] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by Marinovsky**
#### Description

Add `YearStart()` and `YearEnd()` date rules as well as the some unit tests to cover these changes
#### Related Issue

Closes #7988 
#### Motivation and Context

With these changes, users will be able to schedule yearly routines when creating their algorithms in LEAN
#### Requires Documentation Change

N/A
#### How Has This Been Tested?

There were already tests for the monthly date rule so I followed that pattern to make the tests for the yearly date rule
#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Tests/Common/Scheduling/DateRulesTests.cs [hunk 1]
```diff
@@ -313,6 +313,239 @@ public void EndOfMonthWithSymbolWithOffset(Symbols.SymbolsKey symbolKey, int[] e
...
+        {
+            var rules = GetDateRules();
+            var rule = rules.YearStart();
+            var dates = rule.GetDates(new DateTime(2000, 01, 01), new DateTime(2010, 12, 31));
+
+            int count = 0;
+            foreach (var date in dates)
+            {
+                count++;
+                Assert.IsTrue(date.Day <= 4);
```
Comment (jhonabreul) [inline Tests/Common/Scheduling/DateRulesTests.cs:327]: Should this be `Assert.AreEqual(1, date.Day);`? 
Also, you could assert the month as well in all of these tests

File: Tests/Common/Scheduling/DateRulesTests.cs [hunk 2]
```diff
@@ -313,6 +313,239 @@ public void EndOfMonthWithSymbolWithOffset(Symbols.SymbolsKey symbolKey, int[] e
...
+
+            var rules = GetDateRules();
+            var rule = rules.YearStart(10); // 11/1/2000
+            var dates = rule.GetDates(startingDate, endingDate);
+
+            Assert.AreEqual(expectNone, dates.IsNullOrEmpty());
+
+            if (!expectNone)
+            {
+                Assert.AreEqual(new DateTime(2000, 1, 11), dates.First());
```
Comment (jhonabreul) [inline Tests/Common/Scheduling/DateRulesTests.cs:385]: Could be `dates.Single()` just to make sure.
And I guess you could ditch the `expectNone` argument and just check `startingDateDay > 11`.

File: Tests/Common/Scheduling/DateRulesTests.cs [hunk 3]
```diff
@@ -313,6 +313,239 @@ public void EndOfMonthWithSymbolWithOffset(Symbols.SymbolsKey symbolKey, int[] e
...
+            var dates = rule.GetDates(new DateTime(2000, 01, 01), new DateTime(2010, 12, 31));
+
+            int count = 0;
+            foreach (var date in dates)
+            {
+                count++;
+                Assert.AreNotEqual(DayOfWeek.Saturday, date.DayOfWeek);
+                Assert.AreNotEqual(DayOfWeek.Sunday, date.DayOfWeek);
+                Assert.IsTrue(date.Day <= 4);
+                Log.Trace(date.Day.ToString(CultureInfo.InvariantCulture));
```
Comment (jhonabreul) [inline Tests/Common/Scheduling/DateRulesTests.cs:403]: nit - remove or use `Debug`

**Review by jhonabreul — APPROVED** [review]
Looks good! Left a few minor comments for tests

**Review by jhonabreul — APPROVED** [review]
Thank you!

**Review by Martin-Molinero — APPROVED** [review]
Nice work! 👍

**Pull request merged**